sylt-compiler/src/typechecker.rs

```diff
@@ -885,30 +876,62 @@ impl TypeChecker {
 
                 let given_fields: BTreeMap<_, _> = fields
                     .iter()
-                    .map(|(key, expr)| Ok((key.clone(), self.expression(expr, ctx)?)))
+                    .map(|(key, expr)| Ok((key.clone(), self.push_type(Type::Unknown))))
                     .collect::<TypeResult<_>>()?;
 
                 let mut errors = Vec::new();
+                for (field, field_ty) in blob_fields.iter() {
+                    if !given_fields.contains_key(field) {
+                        errors.push(type_error!(
+                            self,
+                            span,
+                            ctx,
+                            TypeError::MissingField {
+                                blob: blob_name.clone(),
+                                field: field.clone(),
+                            }
+                        ));
+                    }
+                }
+
                 for (field, field_ty) in given_fields.iter() {
-                    match blob_fields.get(field) {
-                        Some(_) => {}
-                        None => errors.push(type_error!(
+                    if !blob_fields.contains_key(field) {
+                        errors.push(type_error!(
                             self,
                             span,
                             ctx,
                             TypeError::UnknownField {
                                 blob: blob_name.clone(),
                                 field: field.clone(),
                             }
-                        )),
+                        ));
                     }
                 }
 
                 if !errors.is_empty() {
                     return Err(errors);
                 }
 
-                let given_blob = self.push_type(Type::Blob(blob_name, given_fields));
+                let given_blob =
+                    self.push_type(Type::Blob(blob_name.clone(), given_fields.clone()));
+
+                // Unify the fields with their real types
+                for (key, expr) in fields {
+                    let is_function = matches!(expr.kind, ExpressionKind::Function { .. });
+                    if is_function {
+                        self.stack.push(Variable {
+                            ident: Identifier { name: "self".to_string(), span },
+                            kind: VarKind::Const,
+                            ty: given_blob,
+                        });
+                    }
+                    let expr_ty = self.expression(expr, ctx)?;
+                    self.unify(span, ctx, expr_ty, given_fields[key])?;
+                    if is_function {
+                        self.stack.pop();
+                    }
+                }
```

Comment:
I wonder - sometimes late at night - if the better code truncates the stack to a certain size at the beginning of each loop.

Reply:
No.